fdbserver/include/fdbserver/EncryptKeyProxy.actor.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * EncryptKeyProxy.actor.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// When actually compiled (NO_INTELLISENSE), include the generated version of this file.  In intellisense use the source
+// version.
+#if defined(NO_INTELLISENSE) && !defined(ENCRYPT_KEY_PROXY_ACTOR_G_H)
+#define ENCRYPT_KEY_PROXY_ACTOR_G_H
+#include "fdbserver/EncryptKeyProxy.actor.g.h"
+#elif !defined(ENCRYPT_KEY_PROXY_ACTOR_H)
+#define ENCRYPT_KEY_PROXY_ACTOR_H
+#include "flow/flow.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+namespace EncryptKeyProxy {
+
+inline bool canRetryWith(Error e) {
+	// The below are the only errors that should be retried, all others should throw immediately
+	switch (e.code()) {
+	case error_code_encrypt_keys_fetch_failed:
+	case error_code_timed_out:
+	case error_code_connection_failed:
+		return true;
+	default:
+		return false;
+	}
+}
+
+} // namespace EncryptKeyProxy
+
+ACTOR template <class T>
+Future<T> kmsReqWithExponentialBackoff(std::function<Future<T>()> func, StringRef funcName) {
+	state int numRetries = 0;
+	state double kmsBackoff = FLOW_KNOBS->EKP_KMS_CONNECTION_BACKOFF;
+	TraceEvent("KMSRequestStart").detail("Function", funcName);
+
+	loop {
+		try {
+			T val = wait(func());
+			return val;
+		} catch (Error& e) {
+			TraceEvent(SevWarn, "KMSRequestReceivedError").detail("Function", funcName).detail("ErrorCode", e.code());
+			if (!EncryptKeyProxy::canRetryWith(e)) {
+				throw e;
+			}
+			if (numRetries >= FLOW_KNOBS->EKP_KMS_CONNECTION_RETRIES) {
+				TraceEvent(SevWarnAlways, "KMSRequestRetryLimitExceeded")
+				    .detail("Function", funcName)
+				    .detail("ErrorCode", e.code());
+				// TODO: Should we throw a time out error here?
```

Comment:
One issue with this approach is if `error` changes during the num-retries, the last attempt error gets thrown; however, above we are logging the error which is helpful.

Reply:
Yeah i feel logging the previous errors should be sufficent

Reply:
Am not sure about it as we aren't logging specifics to map what request has failed: for instance: `domainId` and/or `baseCipherId`; just logging functionName may or may-not help much while triaging issue?

One thought is to have caller supply two functions, one defining what needs to be done and second one to generate a more meaningful log on error. Thoughts?

Reply:
Can we also warning log (with suppressFor) on the first retry?

Reply:
> Am not sure about it as we aren't logging specifics to map what request has failed: for instance: domainId and/or baseCipherId

Most if not all of these calls fetch for multiple cipher/domain ids. Not sure exactly what we want to log here

Reply:
We could list all domain ids it's trying to fetch but again im not sure how useful that is

Reply:
Since we retry for key fetch failed, timed out, and connection failures most of these are connection issues so adding more logging probably isnt useful (other than maybe for key fetch failed)

Reply:
EKP and KMS requests have an `optional<UID>` field, which we can use for such cases. The performance penalty is generating a UID per request, but given the operation requires external KMS call, that penalty seems negligible

Reply:
> Since we retry for key fetch failed, timed out, and connection failures most of these are connection issues so adding more logging probably isnt useful (other than maybe for key fetch failed)

Agree, for all above cases, if we go with the callback approach, the logs might be `SevDebug`; still useful while trying to triage things by updating log level on demand. Again, lets think more about it for sure.

Reply:
Sure i can add a callback for custom tracing